### PR TITLE
[Inductor CUTLASS backend] Step 5: Gemm CUTLASS templates

### PR DESCRIPTION
**Pull request by ipiszy**
This is the step 5 to add cutlass as an alternative inductor backend.

Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108015
* #107931
* #107901
* #107847
* #107802

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (Skylion007) [inline torch/_inductor/codegen/cuda/gemm_template.py:365]: Just use `%d% formatting instead of using str()`

File: torch/_inductor/kernel/mm.py
Comment (Skylion007) [inline torch/_inductor/kernel/mm.py:134]:
```suggestion
        log.debug("Added %d cutlass gemm configs.", len(ops))
```

File: torch/_inductor/kernel/mm.py
Comment (Skylion007) [inline torch/_inductor/kernel/mm.py:218]:
```suggestion
        log.debug("Added %d cutlass gemm configs.", len(ops))
```

**Comment by ipiszy** [comment]
Hi @atalman, tests are broken due to cutlass python scripts not installed in the torch package used in unittests. Wonder do you know how to pack cutlass python scripts correctly in a torch package? Thanks!
Example failing test: https://github.com/pytorch/pytorch/actions/runs/5988497553/job/16243956225

File: test/inductor/test_max_autotune.py
Comment (aakhundov) [inline test/inductor/test_max_autotune.py:228]: We don't have control over the generated back-end in the second case, right?

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (aakhundov) [inline torch/_inductor/codegen/cuda/gemm_template.py:82]: Do we need these defined separately, given that each line they're used below has a comment?

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (aakhundov) [inline torch/_inductor/codegen/cuda/gemm_template.py:278]: Should we instead explicitly list the types we support (and return otherwise)? As written, this may break if a new `cutlass_lib.GemmKind` is added.

File: test/inductor/test_max_autotune.py
Comment (ipiszy) [reply test/inductor/test_max_autotune.py:228]: Correct, the final kernel is selected based on autotune result.

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (ipiszy) [reply torch/_inductor/codegen/cuda/gemm_template.py:281]: The existing implementation copies from AIT and is very hacky. It actually relies on GemmKind like Planar, Complex, etc, which doesn't seem make sense. However if I only use GemmKind.Gemm there is no config left. So for now I'd like to keep it as it is. CUTLASS folks suggest using the new Python cutlass interface to generate configs, which might be a follow-up.

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (ipiszy) [reply torch/_inductor/codegen/cuda/gemm_template.py:85]: I feel this has better readability, though it's just a personal choice.

**Review by kadeng — APPROVED** [review]
This looks good to me, assuming
 * This backend is not enabled by default and considered experimental
 * Issues in the previous PRs in the diff stack are adressed before merging
 * The remaining comments in here should be addressed, but even if not, these are minor stylistic issues that should not block the PR in my opinion.
 * CI suite has to pass for the PR to merge

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (aakhundov) [reply torch/_inductor/codegen/cuda/gemm_template.py:297]: I believe, AIT uses an allow-list approach accepting only `GemmKind.Universal` and `GemmKind.Universal3x`: https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/backend/cuda/gemm_universal/common.py#L823-L826 . I guess, this should work here, too? `GemmKind.Gemm` must be something pre-`GemmUniversal`.

Comment (ipiszy) [reply torch/_inductor/codegen/cuda/gemm_template.py]: I see, I tried "GemmKind.Gemm" because we are replacing Gemm with GemmUniversal in generated code. But GemmKind.Gemm doesn't work well. Now I see that AIT actually uses GemmKind.GemmUniversal to emit GemmInstance, which I missed before. Thanks for pointing this out!

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (chenyang78) [inline torch/_inductor/codegen/cuda/gemm_template.py:493]: super nit - I think we may simplify the code to 

```
if len(self.input_nodes) == 2
```

because self.input_nodes[2] will just return None if it's None.

**Review by aakhundov — APPROVED** [review]

**Comment by ipiszy** [comment]
@pytorch merge -f "trunk has 3 failures, unrelated to this PR stack."

**Pull request closed without merging**
